src/Nethermind/Nethermind.TxPool/NonceManager.cs

```diff
@@ -0,0 +1,99 @@
+// SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
+// SPDX-License-Identifier: LGPL-3.0-only
+
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using Nethermind.Core;
+using Nethermind.Int256;
+
+namespace Nethermind.TxPool;
+
+public class NonceManager : INonceManager
+{
+    private readonly ConcurrentDictionary<Address, AddressNonceManager> _addressNonceManagers = new();
+    private readonly IAccountStateProvider _accounts;
+
+    public NonceManager(IAccountStateProvider accounts)
+    {
+        _accounts = accounts;
+    }
+
+    public UInt256 ReserveNonce(Address address)
+    {
+        AddressNonceManager addressNonceManager =
+            _addressNonceManagers.GetOrAdd(address, _ => new AddressNonceManager());
+        return addressNonceManager.ReserveNonce(_accounts.GetAccount(address).Nonce);
+    }
+
+    public void TxAccepted(Address address)
+    {
+        if (_addressNonceManagers.TryGetValue(address, out AddressNonceManager? addressNonceManager))
+        {
+            addressNonceManager.TxAccepted();
+        }
+    }
+
+    public void TxRejected(Address address)
+    {
+        if (_addressNonceManagers.TryGetValue(address, out AddressNonceManager? addressNonceManager))
+        {
+            addressNonceManager.TxRejected();
+        }
+    }
+
+    public void TxWithNonceReceived(Address address, UInt256 nonce)
+    {
+        AddressNonceManager addressNonceManager =
+            _addressNonceManagers.GetOrAdd(address, _ => new AddressNonceManager());
+        addressNonceManager.TxWithNonceReceived(nonce);
+    }
+
+    private class AddressNonceManager
+    {
+        private HashSet<UInt256> _usedNonces = new();
```

Comment:
What a Set brings to the table that we can't achieve with just maxNonce?

Reply:
We will fail in this scenario:
1. sendTx(noce=1)
2. sendTx(nonce=null) nonce=2 will be assigned, but should be 0

Reply:
Got it but is it a viable scenario?

Reply:
Even if user send one tx after another we might receive them in opposite order. For example due to network delays